woocommerce/wc-admin
# Connect the revenue report to the API

This PR is the first step in properly connecting the revenue report up to the API.

The report is not 100% done. Here are a few notes/todos/issues.

* Zero-filled data. After the meeting today, we decided this will be done server side. In in the meantime, if you have gap days, they will not show up on the report. That means some report filters might return an 'empty' response (i.e. if you have 'today' selected with no orders) right now. This can also cause an issue with the 'Previous Period' total on the chart (see screenshot below), because we pull out the matching secondary data point based on the primary dates, and if some are missing from the response, the secondary points don't get added to the chart.
* Different intervals. Right now, you can only view data points as days. More handling needs to be written to support other intervals (hours, weeks, etc).
* No loading indicators or empty state. I have some super temporary unstyled messages implemented at the moment. We still need a full page loading indicator and a message for empty/error states. See #302.
* The data points seem to be off by one day (i.e. the axis shows 29, but if you hover over the point it is for the 30th). This seems to be happening on the hardcoded dashboard example as well. I will log an issue. cc @greenafrican 
* The utilities and code I've written for this report should be useable in other reports. I've left the code in revenue for now while trying to figure everything out, but I expect this code to be moved out so it can be reused by other reports.

Otherwise, you are able to filter the report by date, paginate and sort the table, and view summary numbers.

A note on the API requests: For requests that have results over multiple API pages (i.e. you are viewing a large date span with over 100 results), we need to make requests for all pages so we can properly fill out the graph. I created a utility to help with this.

<img width="1090" alt="screen shot 2018-08-22 at 1 16 33 pm" src="https://user-images.githubusercontent.com/689165/44479154-bd914500-a60d-11e8-9cc8-f8c1a8dcdc2c.png">

<img width="1074" alt="screen shot 2018-08-22 at 1 17 12 pm" src="https://user-images.githubusercontent.com/689165/44479181-d863b980-a60d-11e8-8ff2-efa24fc5d253.png">

To Test:
* run `npm test` and make sure all tests pass
* Before testing on your site, you may want to generate a number of orders. You can use [wc-smooth-generator](https://github.com/woocommerce/wc-smooth-generator) to do this. See the [README](https://github.com/woocommerce/wc-smooth-generator/blob/master/README.md#orders) for an example.
* Pull the latest WooCommerce v3 API from `feature/rest-api-v3-reports` (you might need to deactivate `wc-api-dev`).
* You need to deactivate and reactivate WooCommerce to get the correct summary MySQL tables. There is also a tool under system settings to rebuild the summary tables.
* Visit the revenue report.
* Test reordering columns, paging results, etc.
